Goldtree/Goldtree/Program.cs

```diff
@@ -86,204 +83,228 @@ public static void Initialize()
+            {
+                this.QuitWithError(Messages.USBConnectionError);
+            }
+        }
+
+        private string GetNSPPath()
+        {
+            Log.Log(Messages.NSPSelect);
+            OpenFileDialog fd = new OpenFileDialog()
+            {
+                Title = Messages.NSPFileDialogTitle,
+                Filter = Messages.NSPExtension,
+                Multiselect = false,
+            };
+            if (fd.ShowDialog() != DialogResult.OK)
+            {
+                this.QuitWithError(Messages.NSPSelectError);
+            }
+            return fd.FileName;
+        }
+
+        private void StartInstall()
+        {
             try
             {
                 Command c = new Command(CommandId.ConnectionRequest);
                 USB.Write(c);
-                Log.Log("Attempting to connect to Goldleaf via USB...");
+                Log.Log(Messages.ConnectingToGoldleaf);
                 Command rc = USB.Read();
-                if(rc.MagicOk())
+                if (!rc.MagicOk())
+                {
+                    this.QuitWithError(Messages.USBInvalidCommandReceived);
+                }
+                if (rc.IsCommandId(CommandId.Finish))
+                {
+                    this.QuitWithError(Messages.GoldleafCancelledInstall);
+                }
+                if (!rc.IsCommandId(CommandId.ConnectionResponse))
+                {
+                    this.QuitWithError(Messages.USBInvalidCommandReceived);
+                }
+
+                Log.Log(Messages.GoldleafConnectionEstablished);
+                string nsp = this.GetNSPPath();
+                string nspname = Path.GetFileName(nsp);
+                c = new Command(CommandId.NSPName);
+                USB.Write(c);
+                USB.Write((uint)nspname.Length);
+                USB.Write(nspname);
+                Log.Log(Messages.GoldleafNSPApproval);
+
+                Command rc2 = USB.Read();
+                if (!rc2.MagicOk())
+                {
+                    this.QuitWithError(Messages.USBInvalidCommandReceived);
+                }
+                if (rc2.IsCommandId(CommandId.Finish))
+                {
+                    this.QuitWithError(Messages.GoldleafCancelledInstall);
+                }
+                if (!rc2.IsCommandId(CommandId.Start))
                 {
-                    if(rc.IsCommandId(CommandId.ConnectionResponse))
+                    this.QuitWithError(Messages.USBInvalidCommandReceived);
+                }
+
+                this.SendNSPOverUSB(nsp);    
+            }
+            catch
+            {
+                this.QuitWithError(Messages.NSPGeneralError);
+            }
+        }
+
+        private void SendNSPOverUSB(string nsp)
+        {
+            Log.Log(Messages.GoldleafStartingInstall);
+            try
+            {
+                FileStream fs = new FileStream(nsp, FileMode.Open);
+                StreamStorage ist = new StreamStorage(fs, true);
```

Comment:
FileStream and StreamStorage implement IDisposable and should be wrapped with a set of  `using` blocks.